baseplate/server/__init__.py

```diff
@@ -432,23 +445,90 @@ def load_and_run_shell() -> None:
+            # https://github.com/ipython/ipython/blob/a54bf00feb5182fa821bd5457897b3b30a313436/IPython/core/logger.py#L187-L201
+            f"""
+            ip = get_ipython()
+            from functools import partial
+            ip.magic('logstart {console_logpath}')
+            def log_write(self, data, kind="input"):
+                import time, os
+                if self.log_active and data:
+                    write = self.logfile.write
+                    if kind=='input':
+                        if self.timestamp:
+                            write(time.strftime('# %a, %d %b %Y %H:%M:%S\\n', time.localtime()))
+                        write(f"{{time.strftime('%Y-%m-%d %H:%M:%S', time.localtime())}} {{os.getpid()}} - {{data}}")
+                    elif kind=='output' and self.log_output:
+                        odata = u'\\n'.join([u'#[Out]# %s' % s
+                                        for s in data.splitlines()])
+                        write(u'%s\\n' % odata)
+                    self.logfile.flush()
+            ip.logger.logstop = None
+            ip.logger.log_write = partial(log_write, ip.logger)
+            ip.logger.log_write("Start IPython logging\\n")
+            """
+        ]
         ipython_config.TerminalInteractiveShell.banner2 = banner
+        ipython_config.LoggingMagics.quiet = True
         start_ipython(argv=[], user_ns=env, config=ipython_config)
         raise SystemExit
     except ImportError:
-        import code
-
         newbanner = f"Baseplate Interactive Shell\nPython {sys.version}\n\n"
         banner = newbanner + banner
 
-        # import this just for its side-effects (of enabling nice keyboard
-        # movement while editing text)
         try:
             import readline
 
-            del readline
+            readline.set_completer(Completer(env).complete)
+            readline.parse_and_bind("tab: complete")
+
         except ImportError:
             pass
 
-        shell = code.InteractiveConsole(locals=env)
+        shell = LoggedInteractiveConsole(locals=env, logpath=console_logpath)
         shell.interact(banner)
+
+
+def _get_shell_log_path() -> str:
+    """Determine where to write shell audit logs."""
+    # Define path for console log output
+    if _is_containerized():
+        # write to PID 1 stdout for log aggregation
+        return "/proc/1/fd/1"
+    # otherwise write to a local file
+    return "/var/log/.shell_history"
+
+
+def _is_containerized() -> bool:
+    """Determine if we're running in a container based on cgroup awareness for various container runtimes."""
+    path = "/proc/self/cgroup"
+    in_container = ["kubepods", "docker", "containerd"]
+    if os.path.exists("/.dockerenv"):
+        return True
+    elif os.path.isfile(path):
+        for c in in_container:
+            if (c in line for line in open(path)):
+                return True
```

Comment:
I think this could be a little clearer like this:

```suggestion
    if os.path.exists("/.dockerenv"):
        return True
    
    try:
        with open("/proc/self/cgroup") as my_cgroups_file:
            my_cgroups = my_cgroups_file.read()
            
            for hint in ["kubepods", "docker", "containerd"]:
                if hint in my_cgroups:
                    return True
    except IOError:
        pass
```

A couple of advantages:

* we only read the file in once
* rather than checking if the file exists and then trying to read it as two separate things, we just try to read it and move on if that fails

Reply:
adopted in 60bd099b82bb5149eb64be2629b8b488bd46b290